### PR TITLE
C++ example section

### PR DESCRIPTION
**Pull request by apiaccess21**
C++ inference example with opencv and onnxruntime.

### CODE REVIEW

File: docs/deployment.md [hunk 1]
```diff
@@ -308,3 +308,240 @@ python tools/deploy_test.py \
 - All ONNX and TensorRT models are evaluated with dynamic shapes on the datasets, and images are preprocessed according to the original config file.
 - This tool is still experimental, and we only support a subset of detection and recognition algorithms for now.
 :::
+
+
+## C++ Inference example with OpenCV
+The example below is tested with Visual Studio 2019 as console application, CPU inference only.
+
+**Prerequisites**
```
Comment (gaotongxiao) [inline docs/deployment.md:316]:
```suggestion
### Prerequisites

```

File: docs/deployment.md [hunk 2]
```diff
@@ -308,3 +308,240 @@ python tools/deploy_test.py \
...
+1. Project should use OpenCV (tested with version 4.5.4), ONNX Runtime NuGet package (version 1.9.0).
+2. It's needed to export downloaded (`pth`) Dbnet_r18 detector and Starn_small recognizer models with the following python commands (adjust your paths):
+- `python3.9 ../mmocr/tools/deployment/pytorch2onnx.py --verify --output-file detector.onnx ../mmocr/configs/textdet/dbnet/dbnet_r18_fpnc_1200e_icdar2015.py ./dbnet_r18_fpnc_sbn_1200e_icdar2015_20210329-ba3ab597.pth --dynamic-export det ./sample_big_image_eg_1920x1080.png`
+- `../mmocr/tools/deployment/pytorch2onnx.py --opset 14 --verify --output-file recognizer.onnx ../mmocr/configs/textrecog/satrn/satrn_small.py ./satrn_small_20211009-2cf13355.pth recog ./sample_small_image_eg_200x50.png`
```
Comment (gaotongxiao) [inline docs/deployment.md:320]:
``````suggestion
2. Download *DBNet_r18* detector and *SATRN_small* recognizer models from our [Model Zoo](modelzoo.md), and export them with the following python commands (you may change the paths accordingly):

```bash
python3.9 ../mmocr/tools/deployment/pytorch2onnx.py --verify --output-file detector.onnx ../mmocr/configs/textdet/dbnet/dbnet_r18_fpnc_1200e_icdar2015.py ./dbnet_r18_fpnc_sbn_1200e_icdar2015_20210329-ba3ab597.pth --dynamic-export det ./sample_big_image_eg_1920x1080.png

python3.9 ../mmocr/tools/deployment/pytorch2onnx.py --opset 14 --verify --output-file recognizer.onnx ../mmocr/configs/textrecog/satrn/satrn_small.py ./satrn_small_20211009-2cf13355.pth recog ./sample_small_image_eg_200x50.png
```
``````

File: docs/deployment.md [hunk 3]
```diff
@@ -308,3 +308,240 @@ python tools/deploy_test.py \
...
+> Be aware, while exported `detector.onnx` file is relatively small (about 50 Mb), `recognizer.onnx` is pretty big (more than 600 Mb). Also, Dbnet_r18 can use ONNX opset 11, Satrn_small can be exported with opset 14. 
+
+*Be sure, that verifications of both models are successful - look through the export messages*.
```
Comment (gaotongxiao) [inline docs/deployment.md:323]:
```suggestion

:::{note}
- Be aware, while exported `detector.onnx` file is relatively small (about 50 Mb), `recognizer.onnx` is pretty big (more than 600 Mb).
- *DBNet_r18* can use ONNX opset 11, *SATRN_small* can be exported with opset 14. 
:::

:::{warning}
Be sure, that verifications of both models are successful - look through the export messages.
:::
```

File: docs/deployment.md [hunk 4]
```diff
@@ -308,3 +308,240 @@ python tools/deploy_test.py \
...
+
+3. Example usage of exported models with C++ is in the code below (don't forget to change paths to \*.onnx files). It's applicable to these two models only, other models have another preprocessing and postprocessing logics.
```
Comment (gaotongxiao) [inline docs/deployment.md:325]:
```suggestion
### Example
Example usage of exported models with C++ is in the code below (don't forget to change paths to \*.onnx files). It's applicable to these two models only, other models have another preprocessing and postprocessing logics.
```

File: docs/deployment.md [hunk 5]
```diff
@@ -308,3 +308,240 @@ python tools/deploy_test.py \
...
+		for (int index = 0; index < contours.size(); ++index) {
+			cv::Rect box = expand_box(cv::boundingRect(contours[index]));
+			boxes.push_back(box);
+		}
+
+		return boxes;
+	}
+};
+
+// Satrn_small
```
Comment (gaotongxiao) [inline docs/deployment.md:433]:
```suggestion
// SATRN_small
```

**Review by gaotongxiao — COMMENTED** [review]
Thanks for your contribution! I rephrased some sentences and organize this section in a docs-friendly way. If you feel good about these changes, just apply them and wait for ~3mins before you are able to preview this section in docs (https://mmocr--593.org.readthedocs.build/en/593/deployment.html#c-inference-example-with-opencv).

**Comment by apiaccess21** [comment]
Anaything I need to do?

**Comment by gaotongxiao** [comment]
@apiaccess21 Resolving conversation cannot apply suggested changes to files. You may unresolved them and apply them in batch (https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request)

File: docs/deployment.md (hunk 1 above)
Comment (apiaccess21) [reply docs/deployment.md:316]: Ok, good

File: docs/deployment.md (hunk 2 above)
Comment (apiaccess21) [reply docs/deployment.md:320]: Ok, fine

File: docs/deployment.md (hunk 3 above)
Comment (apiaccess21) [reply docs/deployment.md:323]: Ok, nice :)

File: docs/deployment.md (hunk 4 above)
Comment (apiaccess21) [reply docs/deployment.md:325]: Ok, thanks

File: docs/deployment.md (hunk 5 above)
Comment (apiaccess21) [reply docs/deployment.md:433]: Ok, cool

**Comment by apiaccess21** [comment]
Unfortunately, since the PR was made not from fork (and I have no write access to the repo), I don't see any commit options. Could you resolve and apply everything you have suggested?.. It would be good.

**Comment by gaotongxiao** [comment]
@apiaccess21 Sure, I can push it to your branch. BTW, all I can see is that this PR is created from your fork and I'm not sure what's going wrong.

**Review by gaotongxiao — APPROVED** [review]
Thanks again!

**Comment by apiaccess21** [comment]
Not at all. :) I will share, if I something will be more from my side.

**Comment by fatfishZhao** [comment]
Hi, great job! Do you have any idea how to inference models on GPU with C++?

**Comment by apiaccess21** [comment]
> Hi, great job! Do you have any idea how to inference models on GPU with C++?

Hello! I tried, but had some mess with nvidia drivers. The main difference is in ONNX environment construction changes. Something like next should work, but for now for me it's difficult to test.

```
Detector(const std::string& model_path) {
		Ort::SessionOptions options{ nullptr };
		try {
			OrtCUDAProviderOptions cuda;
			cuda.device_id = 0;
			options.AppendExecutionProvider_CUDA(cuda);
			session = Ort::Session{ env, std::wstring(model_path.begin(), model_path.end()).c_str(), options };
			std::cerr << "Detector uses CUDA" << std::endl;
		} catch (const std::exception& exception) {
			std::cerr << "Detector uses CPU" << std::endl;
			session = Ort::Session{ env, std::wstring(model_path.begin(), model_path.end()).c_str(), Ort::SessionOptions{ nullptr } };
		}		
	}
```

Or, instead, you may use [this sample](https://github.com/microsoft/onnxruntime-inference-examples/blob/18f5702347789218c61729e3252ab53c64865340/c_cxx/imagenet/main.cc#L107).

**Comment by apiaccess21** [comment]
@fatfishZhao , the code above works, with small essential change.
```
Ort::SessionOptions options;
```
That's all.

**Comment by OpenMMLab-Coodinator** [comment]
Hi ！@apiaccess21 , first of all, we want to express our gratitude for your significant PR in the MMOCR project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! **Join us ：https://discord.gg/raweFPmdzG**

If you are Chinese or have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Pull request merged**